zebrad/src/commands/entry_point.rs

```diff
@@ -0,0 +1,138 @@
+
+/// Toplevel entrypoint command.
+///
+/// Handles obtaining toplevel help as well as verbosity settings.
+#[derive(Debug, clap::Parser)]
+#[clap(
+    version = clap::crate_version!(),
+    author="Zcash Foundation <zebra@zfnd.org>",
+    help_template = "\
+{name} {version}\n
+{author}\n
+{usage-heading} {usage}\n
+{all-args}\
+"
+)]
+pub struct EntryPoint {
+    /// Subcommand to execute.
+    ///
+    /// The `command` option will delegate option parsing to the command type,
+    /// starting at the first free argument. Defaults to start.
+    #[clap(subcommand)]
+    pub cmd: Option<ZebradCmd>,
+
+    /// Path to the configuration file
+    #[clap(long, short, help = "path to configuration file")]
+    pub config: Option<PathBuf>,
+
+    /// Increase verbosity setting
+    #[clap(long, short, help = "be verbose")]
+    pub verbose: bool,
+
+    /// Filter strings which override the config file and defaults
+    // This can be applied to the default start command if no subcommand is provided.
+    #[clap(help = "tracing filters which override the zebrad.toml config")]
+    filters: Vec<String>,
+}
+
+impl EntryPoint {
+    /// Borrow the command in the option
+    ///
+    /// # Panics
+    ///
+    /// If `cmd` is None
+    pub fn cmd(&self) -> &ZebradCmd {
+        self.cmd
+            .as_ref()
+            .expect("should default to start if not provided")
+    }
+
+    /// Returns a string that parses to the default subcommand
+    pub fn default_cmd_as_str() -> &'static str {
+        "start"
+    }
+
+    /// Process command arguments and insert the default subcommand
+    /// if no subcommand is provided.
+    pub fn process_cli_args(args: Vec<OsString>) -> Vec<OsString> {
+        // Check if the provided arguments include a subcommand
+        let should_add_default_subcommand = EntryPoint::try_parse_from(&args)
+            .unwrap_or_else(|err| err.exit())
+            .cmd
+            .is_none();
+
+        // Add the default subcommand to args after the top-level args if cmd is None
+        if should_add_default_subcommand {
+            // try_parse_from currently produces an error if the first argument is not the binary name,
+            let mut num_top_level_args = 1;
+
+            // update last_top_level_arg_idx to the number of top-level args
+            for (idx, arg) in args.iter().enumerate() {
+                num_top_level_args = match arg.to_str() {
+                    Some("--verbose" | "-v" | "--version" | "-V" | "--help") => idx + 1,
+                    Some("--config" | "-c") => idx + 2,
+                    _ => num_top_level_args,
+                }
+            }
+
+            let mut updated_args: Vec<_> = args.iter().take(num_top_level_args).cloned().collect();
+            updated_args.push(EntryPoint::default_cmd_as_str().into());
+            updated_args.extend(args.into_iter().skip(num_top_level_args));
+
+            updated_args
+        } else {
+            args
+        }
```

Comment:
Nitpick: this seems like a complicated way to insert an item into a vector

```suggestion
            args.insert(num_top_level_args, EntryPoint::default_cmd_as_str().into());
        }

        args
```